Add Admin Assignment Endpoints to Client – BFF and Authorization Process Integration

Alepazz opened the pull request:
**Jira Link**: https://pagopa.atlassian.net/browse/PIN-6672

## Description
This PR introduces two new endpoints aimed at assigning an ADMIN (i.e., a user) to a client:
- One endpoint in the `authorization-process`
- One corresponding route in the `BFF`

### The implementation raises the following points and questions for discussion:

**Access Control**: Is it correct that only users with the `ADMIN_ROLE` are allowed to call the new endpoint on the _authorization-process_?

**Event Naming**: Is `toCreateEventClientAdminSet` an appropriate and clear name for the event being published?



**Admin Self-Assignment**: A check has been added to detect when the provided adminId matches the current admin of the client. In this case, I decided to return an _HTTP 422 (UNPROCESSABLE_ENTITY)_, under the assumption that replacing an admin with themselves is not a valid use case.This also avoids triggering admin-change events unnecessarily. 
Is it a good choice?
Do we agree that the frontend should prevent selecting the current admin as a replacement?

**Query Result Validation**: Due to the `assertOrganizationIsClientConsumer`, unit tests indicate that the case of an unauthorized user being returned does not occur.
Therefore, `assertOrganizationIsClientConsumer` gives back an array of results, depending on filters BUT, I expect that, using userId as parameter, the result must be at more one. Therefore, an explicit check to ensure that the query returns exactly one user (e.g., if (users.length !== 1)) can be added. Is this acceptable, or should we enforce it regardless?

**Error Responses in BFF**: The current BFF route returns a generic `emptyErrorMessage` for all errors. Should we introduce more specific error messages based on the type of failure returned by the authorization-process?

## ℹ️ Authorization-Updater

This functionality has not been implemented in authorization-updater, as Scala support is being gradually phased out.


### Side Notes
The method `deleteClientKeyByIdErrorMapper` currently does not handle the userNotAllowedOnClient case. Should this be addressed?

AsterITA commented on packages/api-clients/open-api/authorizationApi.yml:
```diff
@@ -430,6 +430,67 @@ paths:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Problem"
+        "422":
```
```suggestion
        "409":
```
IMHO `Conflict` is more appropriate in this case

Alepazz replied:
Right!

AsterITA commented on packages/api-clients/open-api/bffApi.yml:
```diff
@@ -10474,8 +10474,8 @@ paths:
               description: Time interval in milliseconds. Allowed requests will be constantly replenished during the interval. At the end of the interval the max allowed requests will be available
-        "404":
-          description: Client not found
+        "422":
```
```suggestion
        "409":
```
Same as before

AsterITA commented on packages/api-clients/open-api/bffApi.yml:
```diff
@@ -10436,8 +10436,8 @@ paths:
               schema:
                 type: integer
               description: Time interval in milliseconds. Allowed requests will be constantly replenished during the interval. At the end of the interval the max allowed requests will be available
-        "401":
```
This should be kept for when a token can't be validated

AsterITA commented on packages/api-clients/open-api/bffApi.yml:
```diff
@@ -10580,6 +10580,132 @@ paths:
+    post:
+      tags:
+        - clients
+      summary: Adds an Admin to a Client
+      description: Adds an Admin to a Client
+      operationId: addAdminToClient
```
```suggestion
      summary: Sets an Admin to a Client
      description: Sets an Admin to a Client
      operationId: setAdminToClient
```

AsterITA commented on packages/api-clients/open-api/bffApi.yml:
```diff
@@ -10580,6 +10580,132 @@ paths:
+              schema:
+                type: integer
+              description: Time interval in milliseconds. Allowed requests will be constantly replenished during the interval. At the end of the interval the max allowed requests will be available
+        "404":
```
The 409 should be added here too

Alepazz commented:
I've introduced the possibility to call the route for adding an admin to a client within the _authorization-process_, following the discussion [here](https://github.com/pagopa/interop-be-monorepo/pull/1808#pullrequestreview-2787240551).

At the moment, I've chosen to return a `403 status code`. However, both the error code and the overall behavior follow the considerations discussed in the PR above.

**I've also implemented the assertion as middleware using Express.**

Alepazz commented:
⚠️ The merge of this PR is blocked ⚠️ 
as it would be advisable to align with [this development](https://github.com/pagopa/interop-be-monorepo/pull/1754/files) in order to obtain the correct `feature flag` management logic.

AsterITA commented on packages/authorization-platformstate-writer/src/consumerServiceV2.ts:
```diff
@@ -412,6 +414,64 @@ export async function handleMessageV2(
+              kid: key.kid,
+            }),
+            updatedAt: new Date().toISOString(),
+            adminId: unsafeBrandId<UserId>(msg.data.adminId),
```
```suggestion
            adminId: client.adminId,
```
My bad, this should be like this, because `client` is the new one that comes with the event meanwhile `clientEntry` is the old one, so we can correctly get `adminId` from `client`

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
I wouldn't add `adminId` in the old tests since it's an optional field, we can just test it in the new ones

rGregnanin replied:
adminId is set in `getMockClient()` of `testUtils`, so either we remove it from the mock, or we add it to the tests.
I would prefer to add it to the tests since `getMockClient()` also affects other tests.
What do you prefer?

AsterITA replied:
> `adminId` is set in `getMockClient()` of `testUtils`, so either we remove it from the mock, or we add it to the tests. I would prefer to add it to the tests since `getMockClient()` also affects other tests. What do you prefer?

IMHO it shouldn't initialized in `getMockClient()` since it's an optional value and the other tests do not test if its value is correct so it's not useful

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3568,4 +3602,331 @@ describe("integration tests V2 events", async () => {
+          clientId: client.id,
+          kid: key.kid,
+        }),
+        adminId,
```
```suggestion
```
The previous entry didn't have `adminId`, we are adding it with this event, since the event will be skipped this entry shouldn't have `adminId`

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3568,4 +3602,331 @@ describe("integration tests V2 events", async () => {
+
+      const key = getMockKey();
+      const client: Client = {
+        ...getMockClient(),
```
`kind` should be `api`

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3568,4 +3602,331 @@ describe("integration tests V2 events", async () => {
+      expect(retrievedTokenGenStatesEntries).toEqual([tokenClientEntry]);
+    });
+
+    it("ApiClient - should update platform-states entry and insert token-generation-states entry", async () => {
```
I suggest to duplicate this test with an old client which already had an `adminId`

rGregnanin replied:
At this time, as said [here](https://github.com/pagopa/interop-be-monorepo/pull/1804#discussion_r2066259830), the client already has a valued adminId, 
so for me there is no need to duplicate the test, but if you prefer I can do it easily.
What do you think?

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3568,4 +3602,331 @@ describe("integration tests V2 events", async () => {
+
+      const payload: ClientAdminSetV2 = {
+        client: toClientV2(updatedClient),
+        oldAdminId: client.adminId,
```
```suggestion
```
In this case the old client didn't have an `admInId`

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3568,4 +3602,331 @@ describe("integration tests V2 events", async () => {
+        clientPurposesIds: [],
+        clientKind: clientKindToTokenGenerationStatesClientKind(client.kind),
+        clientConsumerId: client.consumerId,
+        clientAdminId: client.adminId,
```
```suggestion
```

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3568,4 +3602,331 @@ describe("integration tests V2 events", async () => {
+          clientId: client.id,
+          kid: oldKey.kid,
+        }),
+        adminId: client.adminId,
```
```suggestion
```

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3705,7 +3777,6 @@ describe("integration tests V2 events", async () => {
 
       const payload: ClientAdminSetV2 = {
         client: toClientV2(updatedClient),
-        oldAdminId: client.adminId,
```
Since this client already had an `admindId`, the payload should have oldAdminId

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3738,7 +3809,7 @@ describe("integration tests V2 events", async () => {
+        kid: client.keys[0].kid,
       });
 
       const tokenClientEntry: TokenGenerationStatesApiClient = {
```
You should duplicate this with the other key, in order to write the tokenGenStates for each key before handling the message, otherwise it seems like we are adding a new key

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3687,15 +3657,117 @@ describe("integration tests V2 events", async () => {
     it("ApiClient - should update platform-states entry and insert token-generation-states entry", async () => {
```
```suggestion
    it("ApiClient - should update platform-states and token-generation-states entries", async () => {
```

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3687,15 +3657,117 @@ describe("integration tests V2 events", async () => {
+      );
+    });
+
+    it("ApiClient - should update platform-states entry and insert token-generation-states entry for a client that already had an adminId", async () => {
```
```suggestion
    it("ApiClient - should update platform-states and token-generation-states entries for a client that already had an adminId", async () => {
```

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -1330,6 +1331,7 @@ describe("integration tests V2 events", async () => {
           kid: addedKey.kid,
         }),
         updatedAt: new Date().toISOString(),
+        adminId: client.adminId,
```
```suggestion
```

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -1406,6 +1408,7 @@ describe("integration tests V2 events", async () => {
           clientId: client.id,
           kid: addedKey.kid,
         }),
+        adminId: client.adminId,
```
```suggestion
```

rGregnanin replied:
good catch 🚀

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -1451,6 +1454,7 @@ describe("integration tests V2 events", async () => {
           kid: addedKey.kid,
         }),
         updatedAt: new Date().toISOString(),
+        adminId: client.adminId,
```
```suggestion
```

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3568,4 +3572,430 @@ describe("integration tests V2 events", async () => {
+        kind: "Api",
+      };
+
+      const adminId: UserId = generateId();
+
+      const payload: ClientAdminSetV2 = {
+        client: toClientV2(client),
+        adminId,
```
```suggestion
      const payload: ClientAdminSetV2 = {
        client: toClientV2(client),
        adminId: generateId(),
```

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3568,4 +3572,430 @@ describe("integration tests V2 events", async () => {
+      );
+    });
+
+    it("ApiClient - should update platform-states entry and update token-generation-states entry", async () => {
```
This test is duplicated, it does what I already suggested to change in the previous tests, we can delete this

AsterITA commented on packages/authorization-platformstate-writer/test/consumerServiceV2.integration.test.ts:
```diff
@@ -3568,4 +3569,361 @@ describe("integration tests V2 events", async () => {
+        adminId: newAdminId,
+      };
+
+      const payload: ClientAdminSetV2 = {
```
This should have `oldAdminId`

AsterITA commented on packages/authorization-process/src/services/authorizationService.ts:
```diff
@@ -566,6 +568,42 @@ export function authorizationServiceBuilder(
+    ): Promise<Client> {
+      assertFeatureFlagEnabled(config, "featureFlagAdminClient");
+      logger.info(`Add or update admin in client ${clientId}`);
+      const client = await retrieveClient(clientId, readModelService);
```
We should add a check for `adminId` which verifies that this `adminId` is indeed an admin for this tenant; you can use `assertUserSelfcareSecurityPrivileges` but with only `ADMIN` as `productRole`

Viktor-K reviewed (approved):
Looks good 👍🏻 
 I leave just a minor considerations

Alepazz commented on packages/authorization-process/src/model/domain/toEvent.ts:
```diff
@@ -133,6 +133,45 @@ export function toCreateEventClientUserAdded(
+  correlationId: CorrelationId,
+  oldAdminId: UserId | undefined
+): CreateEvent<AuthorizationEventV2> {
+  if (oldAdminId) {
+    return {
+      streamId: client.id,
+      version,
+      event: {
+        type: "ClientAdminSet",
+        event_version: 2,
+        data: {
+          client: toClientV2(client),
+          adminId,
+          oldAdminId,
+        },
+      },
+      correlationId,
+    };
+  }
+
+  return {
+    streamId: client.id,
+    version,
+    event: {
+      type: "ClientAdminSet",
+      event_version: 2,
+      data: {
+        client: toClientV2(client),
+        adminId,
+      },
+    },
+    correlationId,
+  };
```
You're right, nice suggestion.
This is something I developed, but since you're working on it, @rGregnanin , will you handle it or should I?

Alepazz replied:
Done.

AsterITA commented on packages/authorization-platformstate-writer/src/consumerServiceV2.ts:
```diff
@@ -412,6 +413,64 @@ export async function handleMessageV2(
+      const pk = makePlatformStatesClientPK(client.id);
+      const clientEntry = await readPlatformClientEntry(pk, dynamoDBClient);
+
+      if (clientEntry && clientEntry.version > msg.version) {
```
As discussed [here](https://github.com/pagopa/interop-be-monorepo/pull/1802#discussion_r2073395717) we should skip only on **>** not **>=**

AsterITA commented:
Remember to add the route to the bruno collection

AsterITA commented on packages/authorization-process/src/utilities/errorMappers.ts:
```diff
@@ -90,6 +90,15 @@ export const addClientUserErrorMapper = (error: ApiError<ErrorCodes>): number =>
+): number =>
+  match(error.code)
+    .with("clientNotFound", () => HTTP_STATUS_NOT_FOUND)
+    .with("organizationNotAllowedOnClient", () => HTTP_STATUS_FORBIDDEN)
```
We should add `userWithoutSecurityPrivileges` and `clientKindNotAllowed`

AsterITA commented on packages/authorization-process/test/addOrUpdateAdminClient.test.ts:
Can you rename the file in `setAdminToClient`?

AsterITA commented on packages/backend-for-frontend/src/routers/clientRouter.ts:
```diff
@@ -182,6 +182,27 @@ const clientRouter = (
+          error,
+          emptyErrorMapper,
+          ctx,
+          `Error adding admin ${req.body.adminId} to client ${req.params.clientId}`
```
```suggestion
          `Error adding user ${req.body.adminId} to client ${req.params.clientId} as admin`
```

AsterITA commented on packages/authorization-process/src/services/authorizationService.ts:
```diff
@@ -567,6 +570,54 @@ export function authorizationServiceBuilder(
+        userRolesToCheck: [userRole.ADMIN_ROLE],
+      });
+
+      logger.info(`Add or update admin in client ${clientId}`);
```
```suggestion
      logger.info(`Set user ${adminId} in client ${clientId} as admin`);
```
I think it's best to have this log before `assertUserSelfcareSecurityPrivileges` since that it's already part of its logic

AsterITA commented on packages/authorization-process/src/services/authorizationService.ts:
```diff
@@ -567,6 +570,55 @@ export function authorizationServiceBuilder(
+          oldAdminId
+        )
+      );
+      return updatedClient;
```
```suggestion
      return { updatedClient, showUsers: true }
```
We should show the users since we already asserted that `OrganizationIsClientConsumer`, I would return this to the router and then use this `showUsers` inside `clientToApiClient`

Viktor-K reviewed (approved):
Great job 💪🏻

ecamellini commented on collections/bff/authorization/Add Client Admin.bru:
```diff
@@ -0,0 +1,26 @@
+
+headers {
+  Authorization: {{JWT}}
+  X-Correlation-Id: {{correlation-id}}
```
BFF does not require correlationId, it generates it

```suggestion
```

ecamellini commented on packages/api-clients/open-api/authorizationApi.yml:
```diff
@@ -430,6 +430,72 @@ paths:
+      - $ref: "#/components/parameters/CorrelationIdHeader"
+      - name: clientId
+        in: path
+        description: ID of Client the users belong to
```
```suggestion
        description: ID of the client where to set the admin user
```
